velox/substrait/VeloxToSubstraitExpr.cpp

```diff
@@ -448,7 +468,38 @@ VeloxToSubstraitExprConvertor::toSubstraitExpr(
   ::substrait::Expression_ReferenceSegment_StructField* directStruct =
       substraitFieldExpr->mutable_direct_reference()->mutable_struct_field();
 
-  directStruct->set_field(inputType->getChildIdx(exprName));
+  // FieldAccessTypedExpr represents one of two things: a leaf in an expression
+  // or a dereference expression(fieldExpr->isInputColumn() == false)
+  // for a leaf in an expression, find idx from child by exprName.
+  // for a dereference expression, find idx from every child by exprName.
+  if (fieldExpr->isInputColumn() ||
+      std::dynamic_pointer_cast<const core::FieldAccessTypedExpr>(
```

Comment:
---- Do we have tests for both code paths ?
Yes, have tests for  both code paths.

* case `(cast(row_constructor(a, b) as row(a bigint, b bigint))).a` is   the case when input of FieldAccessTypedExpr is not FieldAccessTypedExpr.
* case `"(abc).ab.a"` is the normal case.

Here is some context why I add `std::dynamic_pointer_cast<const core::FieldAccessTypedExpr>(fieldExpr->inputs()[0]) == nullptr`. I also think it is not normative. But I do not find a good solution for this coner case.

There are two very similar cases. 
First one is use row_constructor to create a row before `ab.a`.  Second one is put cast and fieldAcess together.
```
  ## First case.
  auto plan =
      PlanBuilder()
          .values({data})
          .project(
              {"cast(row_constructor(a, b) as row(a bigint, b bigint)) as ab"})
          .project(
              {"ab.a"})
          .planNode();

  ## Second case.
  auto plan =
      PlanBuilder()
          .values({data})
          .project(
              {"(cast(row_constructor(a, b) as row(a bigint, b bigint))).a"})
          .planNode();
```
And their substrait plan are the same.
```
 "direct_reference": {
  "struct_field": {
   "field": 0,
   "child": {
    "struct_field": {
     "field": 0
    }
   }
  }
```

But there is a difference for two cases that is the inputType for function `getFieldIdForIntermediateNode`.
First case `inputType->toString()` is  `ROW<ab:ROW<a:BIGINT,b:BIGINT>>`. 
But for the second case, I `inputType->toString()` is "ROW<a:BIGINT,b:INTEGER,c:DOUBLE>". So I will got a null point core in getFieldIdForIntermediateNode function.

So, I think maybe the substrait plan of second case should like this.  
```
 "direct_reference": {
  "struct_field": {
   "field": 0,
   }
  }
```
which means we can ignore "cast" here. because we just want to find `a` from `ROW<a:BIGINT,b:INTEGER,c:DOUBLE>`.  Do you think this is okay? if Ok, we should keep `std::dynamic_pointer_cast<const core::FieldAccessTypedExpr>(fieldExpr->inputs()[0]) == nullptr` logic. @mbasmanova

Reply:
Now I found a better way to fix this condition. [here](https://github.com/facebookincubator/velox/pull/2451/files#diff-65defa30896670ec464dc213f57cf6a8066b5022cf6b862ccc33c332b4a25120R489-R491)

Because only `input of FieldAccessTypedExpr is not FieldAccessTypedExpr` will return struct_field. other condition will not. So it can be consider as find exprName from input directly.  Do you think this is okay？@mbasmanova